src/models/hf/from_hf.rs

```diff
@@ -0,0 +1,55 @@
+use crate::architectures::BuildArchitecture;
+use crate::error::BoxedError;
+use crate::util::renaming_backend::RenamingBackend;
+use candle_core::{DType, Device};
+use candle_nn::var_builder::SimpleBackend;
+use candle_nn::VarBuilder;
+use snafu::{ResultExt, Snafu};
+
+#[derive(Debug, Snafu)]
+pub enum FromHFError {
+    #[snafu(display("Cannot convert Hugging Face model config"))]
+    ConvertConfig { source: BoxedError },
+
+    #[snafu(display("Cannot build model"))]
+    BuildModel { source: BoxedError },
+}
+
+/// Models that can be loaded from Huggingface transformers checkpoints.
+pub trait FromHF {
+    /// Model configuration.
+    type Config: BuildArchitecture<Architecture = Self::Model>
+        + TryFrom<Self::HFConfig, Error = BoxedError>;
+
+    /// HF transformers model configuration.
+    type HFConfig;
+
+    /// The type of model that is constructed.
+    ///
+    /// Note that this is different from `Self`. `Self` is typically a
+    /// unit struct that only implements various loading strategies.
+    /// `Model` is a concrete model type such as `TransformerDecoder`.
+    type Model;
+
+    /// Construct a model from an HF model configuration and parameter backend.
+    ///
+    /// * `hf_config` - The Hugging Face transformers model configuration.
+    /// * `backend` - The parameter store backend.
+    /// * `device` - The device to place the model on.
+    fn from_hf(
+        hf_config: Self::HFConfig,
+        backend: Box<dyn SimpleBackend>,
+        device: Device,
+    ) -> Result<Self::Model, FromHFError> {
+        let config = Self::Config::try_from(hf_config).context(ConvertConfigSnafu)?;
+        let rename_backend = RenamingBackend::new(backend, Self::rename_parameters());
+        let vb = VarBuilder::from_backend(Box::new(rename_backend), DType::F32, device);
```

Comment:
Shouldn't we select the `DType` based on the config?

Reply:
Ah yes, I'll fix this.

Reply:
Fixed. There are several solutions to this:

1. Let the `HFConfig` implement a trait that exposes the data type. This does require all HF config structs to implement an additional trait.
2. Deserialize the data type using a private struct in `FromHFHub`. However, users are then on their own when they use the `FromHF` trait.
3. Provide a small wrapper struct that bolts the data type as an additional field on `HFConfig`. Easy to use for users of the `FromHF` trait, does not require implementing a trait on all HF configuration structs.

For now I opted for (3). This may not prove the best approach in the end, but with the data we have now this seems like the best approach.